src/versions.c

```diff
@@ -126,35 +127,51 @@ void ensure_version_image_exists(int version)
 void write_cookiecrumbs_to_download_area(int version)
 {
 	char *conf;
-	char *cmd;
+	char *filename, *strformat;
+	int versionfd, formatfd;
 
 	conf = config_output_dir();
 	if (conf == NULL) {
 		assert(0);
 	}
 
-	string_or_die(&cmd, "echo %s > %s/%i/swupd-server-src-version", VERSION, conf, version);
-	if (system(cmd) != 0) {
+	string_or_die(&filename, "%s/%i/swupd-server-src-version", conf, version);
+	versionfd = open(filename, O_WRONLY | O_CREAT | O_TRUNC, 0664);
+	if (versionfd == -1) {
 		assert(0);
 	}
-	free(cmd);
+	if (write(versionfd, VERSION, strlen(VERSION)) != strlen(VERSION)) {
```

Comment:
The call to write() can be interrupted, or it may not write all data at first (unlikely in this case, but better to be safe). Or you could use fopen/fwrite instead.